File: experimental/attention/minimax_sparse_attention_split_kv/torch_ops_extension/custom_ops/npu_minimax_sparse_attention_split_kv.py
# -----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
from typing import Optional, Tuple

import torch


def npu_minimax_sparse_attention_split_kv(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    k2q_row_ptr: torch.Tensor,
    k2q_q_indices: torch.Tensor,
    k2q_slot_indices: torch.Tensor,
    actual_seq_lengths: torch.Tensor,
    actual_seq_lengths_kv: torch.Tensor,
    num_key_value_heads: int,
    scale_value: float,
    block_size: int,
    top_k: int,
    *,
    block_table: Optional[torch.Tensor] = None,
    inner_precise: int = 4,
    softmax_lse_flag: bool = False,
    input_layout: str = "TND",
) -> Tuple[torch.Tensor, torch.Tensor]:
    """MiniMax sparse attention split-KV (KV-Gather-Q prefill).

    Encapsulates aclnnMinimaxSparseAttentionSplitKv. Phase1 is KV-centric
    QK → softmax → PV into per-slot partials; Phase2 FlashDecode-combines them.

    Args:
        query (Tensor): Query, bf16 or fp8 e4m3fn. TND [T, N, D], BNSD [B, N, S, D],
            or BSND [B, S, N, D].
        key (Tensor): Key, same dtype as query. Paged
            [num_blocks, block_size, kv_heads, D] (TND only) or contiguous matching
            query layout.
        value (Tensor): Value, same layout/dtype as key.
        k2q_row_ptr (Tensor): int32 CSR row pointers [kv_heads, total_kv_rows+1].
        k2q_q_indices (Tensor): int32 CSR q-token ids [kv_heads, nnz].
            TND uses packed flatten; BNSD/BSND uses padded flatten b*S+t.
        k2q_slot_indices (Tensor): int32 CSR topK slot ids [kv_heads, nnz].
        actual_seq_lengths (Tensor): int32 [B] actual q lengths (0 = dummy request).
        actual_seq_lengths_kv (Tensor): int32 [B] actual kv lengths (0 = dummy request).
        num_key_value_heads (int): KV head count.
        scale_value (float): Softmax scale, typically 1/sqrt(D).
        block_size (int): KV block size (production 128).
        top_k (int): Sparse block budget per q-token.
        block_table (Tensor, optional): int32 [B, max_blocks_per_batch] physical
            block map. Required for paged KV; must be None for BNSD/BSND or
            contiguous TND.
        inner_precise (int): 0 = fp32 softmax + fp32 O_partial; 1 = bf16 softmax
            + bf16 O_partial; 4 (default) = bf16 softmax + fp32 O_partial.
        softmax_lse_flag (bool): If True, write fp32 LSE. TND [T, N, 1],
            BNSD [B, N, S, 1], BSND [B, S, N, 1]. If False, LSE is a [0] placeholder.
        input_layout (str): "TND", "BNSD", or "BSND". Default "TND".
            Paged KV cache requires TND.

    Returns:
        Tuple[Tensor, Tensor]: (attention_out, softmax_lse). attention_out matches
            query shape; dtype is bf16 when query is fp8 e4m3fn, otherwise query dtype.
            softmax_lse is fp32 when softmax_lse_flag is True.

    Example:
        Training people usually have indexer ``select_idx``, not CSR. Build CSR
        first, then call this op. Typical BNSD contiguous (no paged cache)::

            import custom_ops
            from custom_ops import build_k2q_csr, npu_minimax_sparse_attention_split_kv

            row_ptr, q_idx, slot_idx = build_k2q_csr(
                select_idx, actual_seq_lengths, actual_seq_lengths_kv, block_size,
                input_layout="BNSD",
            )
            attn_out, softmax_lse = npu_minimax_sparse_attention_split_kv(
                query, key, value, row_ptr, q_idx, slot_idx,
                actual_seq_lengths, actual_seq_lengths_kv,
                num_key_value_heads, scale_value, block_size, top_k,
                input_layout="BNSD",
            )

    Note:
        Forward-only prefill kernel. No autograd / backward is registered.
        ``group_size = num_q_heads / num_key_value_heads`` must be in ``[1, 16]``,
        and head dim must be 128.
    """
    return torch.ops.custom.npu_minimax_sparse_attention_split_kv(
        query,
        key,
        value,
        block_table,
        k2q_row_ptr,
        k2q_q_indices,
        k2q_slot_indices,
        actual_seq_lengths,
        actual_seq_lengths_kv,
        num_key_value_heads,
        scale_value,
        block_size,
        top_k,
        inner_precise,
        softmax_lse_flag,
        input_layout,
    )

File: experimental/attention/minimax_sparse_attention_split_kv/torch_ops_extension/custom_ops/csrc/npu_minimax_sparse_attention_split_kv.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file npu_minimax_sparse_attention_split_kv.cpp
 * \brief MiniMaxSparseAttentionSplitKv PyTorch NPU extension
 */

#include <torch/library.h>
#include <c10/core/DeviceGuard.h>
#include <string>
#include "ops_common.h"

namespace custom {
using namespace at_npu::native;

constexpr int64_t DIM_THREE = 3;
constexpr int64_t DIM_FOUR = 4;

void CheckInt32Tensor(const at::Tensor &tensor, const char *name)
{
    TORCH_CHECK(tensor.scalar_type() == at::kInt, name, " dtype must be int32, got ", tensor.scalar_type());
}

static void CheckInputs(const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
                        const c10::optional<at::Tensor> &blockTable, const at::Tensor &k2qRowPtr,
                        const at::Tensor &k2qQIndices, const at::Tensor &k2qSlotIndices,
                        const at::Tensor &actualSeqLengths, const at::Tensor &actualSeqLengthsKv,
                        int64_t numKeyValueHeads, int64_t blockSize, int64_t topK, int64_t innerPrecise,
                        const std::string &inputLayout)
{
    TORCH_CHECK(query.numel() > 0, "Tensor query is empty.");
    TORCH_CHECK(key.numel() > 0, "Tensor key is empty.");
    TORCH_CHECK(value.numel() > 0, "Tensor value is empty.");
    TORCH_CHECK(innerPrecise == 0 || innerPrecise == 1 || innerPrecise == 4, "inner_precise must be 0, 1 or 4, got ",
                innerPrecise);
    TORCH_CHECK(inputLayout == "TND" || inputLayout == "BNSD" || inputLayout == "BSND",
                "input_layout must be TND, BNSD or BSND, got ", inputLayout);
    TORCH_CHECK(numKeyValueHeads > 0, "num_key_value_heads must be > 0, got ", numKeyValueHeads);
    TORCH_CHECK(blockSize > 0, "block_size must be > 0, got ", blockSize);
    TORCH_CHECK(topK > 0, "top_k must be > 0, got ", topK);
    if (blockTable.has_value() && blockTable.value().defined()) {
        CheckInt32Tensor(blockTable.value(), "block_table");
    }
    CheckInt32Tensor(k2qRowPtr, "k2q_row_ptr");
    CheckInt32Tensor(k2qQIndices, "k2q_q_indices");
    CheckInt32Tensor(k2qSlotIndices, "k2q_slot_indices");
    CheckInt32Tensor(actualSeqLengths, "actual_seq_lengths");
    CheckInt32Tensor(actualSeqLengthsKv, "actual_seq_lengths_kv");
    if (inputLayout == "TND") {
        TORCH_CHECK(query.dim() == DIM_THREE, "TND query must be rank 3 [T, N, D], got ", query.dim());
    } else {
        TORCH_CHECK(query.dim() == DIM_FOUR, inputLayout, " query must be rank 4, got ", query.dim());
    }
}

static c10::TensorOptions AttentionOutOptions(const at::Tensor &query)
{
    auto outOpts = query.options();
    // FP8 Q/K/V still writes BF16 attentionOut (see infershape / aclnn).
    if (query.scalar_type() == at::kFloat8_e4m3fn) {
        outOpts = outOpts.dtype(at::kBFloat16);
    }
    return outOpts;
}

static std::tuple<at::Tensor, at::Tensor> ConstructOutputs(const at::Tensor &query, bool softmaxLseFlag,
                                                           const std::string &inputLayout)
{
    at::Tensor attentionOut = at::empty(query.sizes(), AttentionOutOptions(query));
    at::Tensor softmaxLse;
    if (softmaxLseFlag) {
        if (inputLayout == "TND") {
            softmaxLse = at::empty({query.size(0), query.size(1), 1}, query.options().dtype(at::kFloat));
        } else {
            softmaxLse = at::empty({query.size(0), query.size(1), query.size(2), 1}, query.options().dtype(at::kFloat));
        }
    } else {
        softmaxLse = at::empty({0}, query.options().dtype(at::kFloat));
    }
    return std::tuple<at::Tensor, at::Tensor>(attentionOut, softmaxLse);
}

std::tuple<at::Tensor, at::Tensor> npu_minimax_sparse_attention_split_kv_npu(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const c10::optional<at::Tensor> &blockTable, const at::Tensor &k2qRowPtr, const at::Tensor &k2qQIndices,
    const at::Tensor &k2qSlotIndices, const at::Tensor &actualSeqLengths, const at::Tensor &actualSeqLengthsKv,
    int64_t numKeyValueHeads, double scaleValue, int64_t blockSize, int64_t topK, int64_t innerPrecise,
    bool softmaxLseFlag, c10::string_view inputLayout)
{
    std::string inputLayoutStr(inputLayout);
    CheckInputs(query, key, value, blockTable, k2qRowPtr, k2qQIndices, k2qSlotIndices, actualSeqLengths,
                actualSeqLengthsKv, numKeyValueHeads, blockSize, topK, innerPrecise, inputLayoutStr);

    at::Tensor attentionOut;
    at::Tensor softmaxLse;
    {
        const c10::OptionalDeviceGuard deviceGuard(c10::Device(query.device()));
        // Zero-init so BNSD/BSND padding tokens (kernel leaves them unwritten) stay 0.
        attentionOut = at::zeros(query.sizes(), AttentionOutOptions(query));
        if (softmaxLseFlag) {
            if (inputLayoutStr == "TND") {
                softmaxLse = at::zeros({query.size(0), query.size(1), 1}, query.options().dtype(at::kFloat));
            } else {
                softmaxLse =
                    at::zeros({query.size(0), query.size(1), query.size(2), 1}, query.options().dtype(at::kFloat));
            }
        } else {
            softmaxLse = at::empty({0}, query.options().dtype(at::kFloat));
        }
    }

    char *layoutPtr = const_cast<char *>(inputLayoutStr.c_str());
    EXEC_NPU_CMD_V1(aclnnMinimaxSparseAttentionSplitKv, query, key, value, blockTable, k2qRowPtr, k2qQIndices,
                    k2qSlotIndices, actualSeqLengths, actualSeqLengthsKv, numKeyValueHeads, scaleValue, blockSize, topK,
                    innerPrecise, softmaxLseFlag, layoutPtr, attentionOut, softmaxLse);
    return std::tuple<at::Tensor, at::Tensor>(attentionOut, softmaxLse);
}

std::tuple<at::Tensor, at::Tensor> npu_minimax_sparse_attention_split_kv_meta(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const c10::optional<at::Tensor> &blockTable, const at::Tensor &k2qRowPtr, const at::Tensor &k2qQIndices,
    const at::Tensor &k2qSlotIndices, const at::Tensor &actualSeqLengths, const at::Tensor &actualSeqLengthsKv,
    int64_t numKeyValueHeads, double, int64_t blockSize, int64_t topK, int64_t innerPrecise, bool softmaxLseFlag,
    c10::string_view inputLayout)
{
    std::string inputLayoutStr(inputLayout);
    CheckInputs(query, key, value, blockTable, k2qRowPtr, k2qQIndices, k2qSlotIndices, actualSeqLengths,
                actualSeqLengthsKv, numKeyValueHeads, blockSize, topK, innerPrecise, inputLayoutStr);
    return ConstructOutputs(query, softmaxLseFlag, inputLayoutStr);
}

} // namespace custom

TORCH_LIBRARY_IMPL(custom, PrivateUse1, m)
{
    m.impl("npu_minimax_sparse_attention_split_kv", &custom::npu_minimax_sparse_attention_split_kv_npu);
}

TORCH_LIBRARY_IMPL(custom, Meta, m)
{
    m.impl("npu_minimax_sparse_attention_split_kv", &custom::npu_minimax_sparse_attention_split_kv_meta);
}
